File: backend/models/Patient.js
const mongoose = require("mongoose");

const patientSchema = new mongoose.Schema({
  patientId: {
    type: String,
    required: true,
    unique: true,
  },
  name: {
    type: String,
    required: true,
  },
  email: {
    type: String,
    required: true,
    unique: true,
  },
  password: {
    type: String,
    required: true,
  },
  phone: {
    type: String,
    required: true,
  },
  age: {
    type: Number,
    required: true,
    min: 0,
    max: 150,
  },
  gender: {
    type: String,
    enum: ["male", "female", "other"],
    required: true,
  },
  dateOfBirth: {
    type: Date,
    required: true,
  },
  address: {
    street: String,
    city: String,
    state: String,
    pincode: String,
    country: {
      type: String,
      default: "India",
    },
  },
  emergencyContacts: [
    {
      name: {
        type: String,
        required: true,
      },
      relationship: {
        type: String,
        required: true,
      },
      phone: {
        type: String,
        required: true,
      },
      isPrimary: {
        type: Boolean,
        default: false,
      },
    },
  ],
  medicalHistory: [
    {
      condition: {
        type: String,
        required: true,
      },
      diagnosedDate: Date,
      notes: String,
      doctor: String,
      status: {
        type: String,
        enum: ["active", "resolved", "chronic"],
        default: "active",
      },
    },
  ],
  allergies: [
    {
      allergen: String,
      severity: {
        type: String,
        enum: ["mild", "moderate", "severe"],
        default: "mild",
      },
      reaction: String,
    },
  ],
  currentMedications: [
    {
      name: String,
      dosage: String,
      frequency: String,
      prescribedBy: String,
      startDate: Date,
      endDate: Date,
    },
  ],
  vitals: {
    bloodPressure: {
      systolic: Number,
      diastolic: Number,
      lastUpdated: Date,
    },
    heartRate: {
      value: Number,
      lastUpdated: Date,
    },
    temperature: {
      value: Number,
      unit: {
        type: String,
        enum: ["celsius", "fahrenheit"],
        default: "celsius",
      },
      lastUpdated: Date,
    },
    weight: {
      value: Number,
      unit: {
        type: String,
        enum: ["kg", "lbs"],
        default: "kg",
      },
      lastUpdated: Date,
    },
    height: {
      value: Number,
      unit: {
        type: String,
        enum: ["cm", "ft"],
        default: "cm",
      },
      lastUpdated: Date,
    },
  },
  preferredLanguage: {
    type: String,
    enum: ["en", "hi", "pa"],
    default: "en",
  },
  bloodGroup: {
    type: String,
    enum: ["A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-"],
  },
  profileImage: String,
  isActive: {
    type: Boolean,
    default: true,
  },
  lastConsultation: {
    type: Date,
  },
  totalConsultations: {
    type: Number,
    default: 0,
  },
  insuranceInfo: {
    provider: String,
    policyNumber: String,
    validUntil: Date,
  },
  notifications: {
    email: {
      type: Boolean,
      default: true,
    },
    sms: {
      type: Boolean,
      default: true,
    },
    push: {
      type: Boolean,
      default: true,
    },
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
  updatedAt: {
    type: Date,
    default: Date.now,
  },
});

// Indexes for better performance
patientSchema.index({ email: 1 });
patientSchema.index({ patientId: 1 });
patientSchema.index({ phone: 1 });

// Pre-save middleware
patientSchema.pre("save", function (next) {
  this.updatedAt = Date.now();
  next();
});

// Virtual for BMI calculation
patientSchema.virtual("bmi").get(function () {
  if (this.vitals.weight?.value && this.vitals.height?.value) {
    const weightInKg =
      this.vitals.weight.unit === "lbs"
        ? this.vitals.weight.value * 0.453592
        : this.vitals.weight.value;

    const heightInM =
      this.vitals.height.unit === "ft"
        ? this.vitals.height.value * 0.3048
        : this.vitals.height.value / 100;

    return (weightInKg / (heightInM * heightInM)).toFixed(1);
  }
  return null;
});

// Virtual for full address
patientSchema.virtual("fullAddress").get(function () {
  if (!this.address) return "";
  return `${this.address.street}, ${this.address.city}, ${this.address.state} ${this.address.pincode}`;
});

// Method to add medical history
patientSchema.methods.addMedicalHistory = function (condition, notes, doctor) {
  this.medicalHistory.push({
    condition,
    notes,
    doctor,
    diagnosedDate: new Date(),
  });
  return this.save();
};

// Method to update vitals
patientSchema.methods.updateVitals = function (vitalsData) {
  Object.keys(vitalsData).forEach((key) => {
    if (this.vitals[key]) {
      this.vitals[key].value = vitalsData[key].value;
      this.vitals[key].lastUpdated = new Date();
      if (vitalsData[key].unit) {
        this.vitals[key].unit = vitalsData[key].unit;
      }
    }
  });
  return this.save();
};

module.exports = mongoose.model("Patient", patientSchema);

File: admin_panel/lib/models/patient.dart
class Patient {
  final String id;
  final String name;
  final String email;
  final String phone;
  final int age;
  final String gender;
  final String? patientId;
  final String? address;
  final String? emergencyContact;
  final String? medicalHistory;
  final List<String> allergies;
  final String? bloodGroup;
  final DateTime createdAt;
  final DateTime? lastVisit;
  final bool isActive;
  final String? profileImage;
  final List<String> medications;
  final List<String> medicalRecords;

  Patient({
    required this.id,
    required this.name,
    required this.email,
    required this.phone,
    required this.age,
    required this.gender,
    this.patientId,
    this.address,
    this.emergencyContact,
    this.medicalHistory,
    this.allergies = const [],
    this.bloodGroup,
    required this.createdAt,
    this.lastVisit,
    this.isActive = true,
    this.profileImage,
    this.medications = const [],
    this.medicalRecords = const [],
  });

  factory Patient.fromJson(Map<String, dynamic> json) {
    return Patient(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? '',
      email: json['email'] ?? '',
      phone: json['phone'] ?? '',
      age: json['age'] ?? 0,
      gender: json['gender'] ?? '',
      patientId: json['patientId'],
      address: json['address'],
      emergencyContact: json['emergencyContact'],
      medicalHistory: json['medicalHistory'],
      allergies:
          json['allergies'] != null ? List<String>.from(json['allergies']) : [],
      bloodGroup: json['bloodGroup'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      lastVisit:
          json['lastVisit'] != null ? DateTime.parse(json['lastVisit']) : null,
      isActive: json['isActive'] ?? true,
      profileImage: json['profileImage'],
      medications: json['medications'] != null
          ? List<String>.from(json['medications'])
          : [],
      medicalRecords: json['medicalRecords'] != null
          ? List<String>.from(json['medicalRecords'])
          : [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'email': email,
      'phone': phone,
      'age': age,
      'gender': gender,
      'patientId': patientId,
      'address': address,
      'emergencyContact': emergencyContact,
      'medicalHistory': medicalHistory,
      'allergies': allergies,
      'bloodGroup': bloodGroup,
      'createdAt': createdAt.toIso8601String(),
      'lastVisit': lastVisit?.toIso8601String(),
      'isActive': isActive,
      'profileImage': profileImage,
      'medications': medications,
      'medicalRecords': medicalRecords,
    };
  }

  Patient copyWith({
    String? id,
    String? name,
    String? email,
    String? phone,
    int? age,
    String? gender,
    String? patientId,
    String? address,
    String? emergencyContact,
    String? medicalHistory,
    List<String>? allergies,
    String? bloodGroup,
    DateTime? createdAt,
    DateTime? lastVisit,
    bool? isActive,
    String? profileImage,
    List<String>? medications,
    List<String>? medicalRecords,
  }) {
    return Patient(
      id: id ?? this.id,
      name: name ?? this.name,
      email: email ?? this.email,
      phone: phone ?? this.phone,
      age: age ?? this.age,
      gender: gender ?? this.gender,
      patientId: patientId ?? this.patientId,
      address: address ?? this.address,
      emergencyContact: emergencyContact ?? this.emergencyContact,
      medicalHistory: medicalHistory ?? this.medicalHistory,
      allergies: allergies ?? this.allergies,
      bloodGroup: bloodGroup ?? this.bloodGroup,
      createdAt: createdAt ?? this.createdAt,
      lastVisit: lastVisit ?? this.lastVisit,
      isActive: isActive ?? this.isActive,
      profileImage: profileImage ?? this.profileImage,
      medications: medications ?? this.medications,
      medicalRecords: medicalRecords ?? this.medicalRecords,
    );
  }
}
